packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -661,6 +661,8 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
      * @param requestHandler - Request handlers for the container runtime
      * @param runtimeOptions - Additional options to be passed to the runtime
      * @param existing - (optional) When loading from an existing snapshot. Precedes context.existing if provided
+     * @param containerRuntimeCtor - (optional) Constructor to use to create the ContainerRuntime instance. This
+     * allows mixin classes to leverage this method to define their own async initializer.
      */
     public static async load(
```

Comment:
Is there a way for this API to be a stand-alone function?
I think it makes more sense, given that it may be creating different type.

Alternatively, we can break initialization into sync constructor, and async init() function that has to be called after creation, and equivalent of ContainerRuntime.load() becomes something like

```ts
class A extends ContainerRuntime {}

async function load() {
   const a = new A();
   await a.init();
   return a;
}
```
It's likely less appealing, as many properties on class would need to become optional, and a lot of code needs to use ! or checks.

Reply:
There was some discussion on this on [this thread](https://github.com/microsoft/FluidFramework/pull/13073#discussion_r1036383769). My main desire here was to not expose constructor visibility any further. I'm not very strongly opinionated on this area if you have alternative structures that untangle things. You can also see that the implemented pattern causes mixins which add state to the class to have a similar problem to your suggestion--augmenting the state in a static initializer introduces potential undefinedness.